### PR TITLE
Fix frame dropping in filtergraph at EOF

### PR DESCRIPTION
**Pull request by j0sh**
ffmpeg: Fix frame dropping in filtergraph at EOF.
    
This is done by setting the EOF timestamp of the filtergraph
such that the timestamp is *after* the last frame inserted.
    
Also add tests exhibiting various cases uncovered in the process
of fixing this.

Fixes https://github.com/livepeer/lpms/issues/154

### CODE REVIEW

File: ffmpeg/ffmpeg_test.go [hunk 1]
```diff
@@ -1037,3 +1048,292 @@ func TestTranscoder_StreamCopyAndDrop(t *testing.T) {
...
+	in.Fname = dir + "/out.mp4"
+	res2, err := Transcode3(in, nil)
+	if err != nil {
+		t.Error(err)
+	}
+	// TODO Ideally these two should match. As far as I can tell it is due
+	//      to timestamp rounding around EOF. Note this does not happen with
+	//      mpegts formatted output!
+	if res2.Decoded.Frames != 60 || res.Encoded[0].Frames != 61 {
+		t.Error("Did not get expected frame counts: is this fixed?!? ",
```
Comment (mkrufky) [inline ffmpeg/ffmpeg_test.go]: do we need the "is this fixed?!?" bit?

**Comment by j0sh** [comment]
> I believe that the same affect achieved by this PR would be achieved by the pre-pending and post-pending of one audio frame

This PR doesn't attempt to address the audio popping issue; it only fixes frame dropping from the filtergraph as described in https://github.com/livepeer/lpms/issues/154 . The audio issue (along with your proposed fix) is still open and documented in https://github.com/livepeer/lpms/issues/152 .

File: ffmpeg/ffmpeg_test.go (hunk 1 above)
Comment (j0sh) [reply ffmpeg/ffmpeg_test.go]: Happy to make adjustments to the language if you can think of a better phrasing :) Mostly, the intent was to express some notion of:

* This behavior is currently broken, and this unit test serves as reproducible documentation (along with https://github.com/livepeer/lpms/issues/155)
* But we need to make broken tests "pass"
* If the broken test no longer passes, maybe it means the underlying issue has been fixed? 🤞 
* If the test breaks, hopefully a distinct error message will prompt more careful review of the issue and any associated changes.

Comment (mkrufky) [reply ffmpeg/ffmpeg_test.go]: ok, that makes a lot more sense now - thank you for explaining :-)

**Comment by darkdarkdragon** [comment]
@j0sh can't find way to comment on code that is not changed, so asking here:
here:
```c
  if (!filter || !filter->active) {
    // No filter in between decoder and encoder, so use input frame directly
    ret = encode(encoder, inf, octx, ost);
    if (ret < 0) return ret;
  }
```

shouldn't it exit unconditionally, like this:
```c
  if (!filter || !filter->active) {
    // No filter in between decoder and encoder, so use input frame directly
    ret = encode(encoder, inf, octx, ost);
    return ret;
  }
```

File: ffmpeg/ffmpeg_test.go (hunk 1 above)
Comment (j0sh) [reply ffmpeg/ffmpeg_test.go]: Updated the language a bit in https://github.com/livepeer/lpms/pull/153/commits/ab98e5177bf328322e594a3efdc86a0dc3225ab2

**Comment by j0sh** [comment]
> shouldn't it exit unconditionally, like this:

@darkdarkdragon Ah yes indeed! Fixed in https://github.com/livepeer/lpms/pull/153/commits/ab98e5177bf328322e594a3efdc86a0dc3225ab2

(Note that we don't actually hit this code path in LPMS itself but I've used it in the past for testing the C API directly...)

**Pull request merged**